File: tests/domains/media/test_download_sfx.py
from typing import TYPE_CHECKING

from core.errors import ToolError
from core.freesound import Freesound, SoundCandidate
from domains import Runtime
from domains.media.download_sfx import download_sound_effect
from tests.helpers import unwrap

if TYPE_CHECKING:
    from pathlib import Path

    import pytest

BOOM = SoundCandidate(
    sound_id=785925,
    name="Drama Boom",
    duration=4.4,
    license="CC0",
    author="modusmogulus",
    tags=["boom"],
    preview_url="https://cdn.freesound.org/previews/785/785925-hq.mp3",
)


def _fake_freesound(monkeypatch: pytest.MonkeyPatch) -> None:
    def sound(_self: Freesound, sound_id: int) -> SoundCandidate:
        if sound_id != BOOM["sound_id"]:
            raise ToolError("Som não encontrado no Freesound.", code="not_found")
        return BOOM

    def download(_self: Freesound, candidate: SoundCandidate, target_dir: Path) -> Path:
        target_dir.mkdir(parents=True, exist_ok=True)
        target = target_dir / f"Drama_Boom_{candidate['sound_id']}.mp3"
        target.write_bytes(b"mp3")
        return target

    monkeypatch.setattr(Freesound, "sound", sound)
    monkeypatch.setattr(Freesound, "download_preview", download)


def test_download_success(runtime: Runtime, monkeypatch: pytest.MonkeyPatch) -> None:
    _fake_freesound(monkeypatch)
    result = unwrap(download_sound_effect(runtime, 785925))
    assert result["output"] == "sfx/Drama_Boom_785925.mp3"
    assert result["name"] == "Drama Boom"
    assert result["license"] == "CC0"
    assert (runtime.workspace.root / "sfx" / "Drama_Boom_785925.mp3").is_file()


def test_download_custom_folder(runtime: Runtime, monkeypatch: pytest.MonkeyPatch) -> None:
    _fake_freesound(monkeypatch)
    result = unwrap(download_sound_effect(runtime, 785925, folder="sons/memes"))
    assert result["output"] == "sons/memes/Drama_Boom_785925.mp3"


def test_download_not_found(runtime: Runtime, monkeypatch: pytest.MonkeyPatch) -> None:
    _fake_freesound(monkeypatch)
    assert download_sound_effect(runtime, 1).get("code") == "not_found"


def test_download_invalid_id(runtime: Runtime) -> None:
    assert download_sound_effect(runtime, -5).get("code") == "invalid_argument"


def test_download_folder_outside_workspace(runtime: Runtime) -> None:
    assert download_sound_effect(runtime, 785925, folder="../fora").get("code") == (
        "outside_workspace"
    )


def test_download_folder_is_file(runtime: Runtime, monkeypatch: pytest.MonkeyPatch) -> None:
    _fake_freesound(monkeypatch)
    (runtime.workspace.root / "arquivo.txt").write_text("x")
    assert download_sound_effect(runtime, 785925, folder="arquivo.txt").get("code") == (
        "invalid_argument"
    )

File: tests/domains/media/test_download_sfx.rs
//! Testes de `domains/media/download_sfx.rs`, com o transporte do Freesound substituído.

use std::collections::HashMap;
use std::sync::Arc;

use mcp_tools::config::Settings;
use mcp_tools::core::downloader::Downloader;
use mcp_tools::core::errors::ErrorCode;
use mcp_tools::core::ffmpeg::FFmpeg;
use mcp_tools::core::freesound::{http_error, Freesound};
use mcp_tools::domains::media::download_sfx::download_sound_effect;
use mcp_tools::domains::Runtime;
use serde_json::json;

use crate::conftest::{jobs, workspace, TestRuntime};

const BOOM_ID: i64 = 785925;
const PREVIEW: &str = "https://cdn.freesound.org/previews/785/785925-hq.mp3";

/// Runtime cujo Freesound conhece só o som 785925 ("Drama Boom") e entrega
/// bytes fixos no preview. Qualquer outro id responde 404.
fn runtime_with_freesound() -> TestRuntime {
    let freesound = Freesound::with_fetcher(
        "abc",
        Arc::new(|url: &str| {
            if url == PREVIEW {
                return Ok(b"mp3".to_vec());
            }
            if url.starts_with(&format!("https://freesound.org/apiv2/sounds/{BOOM_ID}/")) {
                let body = json!({
                    "id": BOOM_ID,
                    "name": "Drama Boom",
                    "duration": 4.4,
                    "license": "http://creativecommons.org/publicdomain/zero/1.0/",
                    "username": "modusmogulus",
                    "tags": ["boom"],
                    "previews": {"preview-hq-mp3": PREVIEW},
                });
                return Ok(serde_json::to_vec(&body).unwrap());
            }
            Err(http_error(404))
        }),
    );
    let ws = workspace();
    let mut env = HashMap::new();
    env.insert(
        "WORKSPACE_DIR".to_string(),
        ws.workspace.root.to_string_lossy().into_owned(),
    );
    let runtime = Runtime {
        settings: Settings::from_map(&env).expect("settings"),
        workspace: ws.workspace,
        ffmpeg: FFmpeg::default(),
        jobs: jobs(),
        downloader: Downloader::default(),
        freesound,
    };
    TestRuntime {
        dir: ws.dir,
        runtime: Arc::new(runtime),
    }
}

#[test]
fn test_download_success() {
    let rt = runtime_with_freesound();
    let result = download_sound_effect(&rt.runtime, BOOM_ID, "sfx").unwrap();
    assert_eq!(result.output, "sfx/Drama_Boom_785925.mp3");
    assert_eq!(result.name, "Drama Boom");
    assert_eq!(result.license, "CC0");
    assert_eq!(result.author, "modusmogulus");
    assert!(rt
        .root()
        .join("sfx")
        .join("Drama_Boom_785925.mp3")
        .is_file());
}

#[test]
fn test_download_custom_folder() {
    let rt = runtime_with_freesound();
    let result = download_sound_effect(&rt.runtime, BOOM_ID, "sons/memes").unwrap();
    assert_eq!(result.output, "sons/memes/Drama_Boom_785925.mp3");
}

#[test]
fn test_download_not_found() {
    let rt = runtime_with_freesound();
    let err = download_sound_effect(&rt.runtime, 1, "sfx").unwrap_err();
    assert_eq!(err.code, ErrorCode::NotFound);
}

#[test]
fn test_download_invalid_id() {
    let rt = runtime_with_freesound();
    let err = download_sound_effect(&rt.runtime, -5, "sfx").unwrap_err();
    assert_eq!(err.code, ErrorCode::InvalidArgument);
}

#[test]
fn test_download_folder_outside_workspace() {
    let rt = runtime_with_freesound();
    let err = download_sound_effect(&rt.runtime, BOOM_ID, "../fora").unwrap_err();
    assert_eq!(err.code, ErrorCode::OutsideWorkspace);
}

#[test]
fn test_download_folder_is_file() {
    let rt = runtime_with_freesound();
    std::fs::write(rt.root().join("arquivo.txt"), "x").unwrap();
    let err = download_sound_effect(&rt.runtime, BOOM_ID, "arquivo.txt").unwrap_err();
    assert_eq!(err.code, ErrorCode::InvalidArgument);
}
